number_theory/modular.h

```diff
@@ -111,6 +112,120 @@ class Modular {
   }
 };
 
+namespace modular_internal {
+
+// Tests if the type T is the Modular class.
+template <typename T>
+struct IsModularImpl {
+  template <auto mod,
+            std::enable_if_t<std::is_same_v<T, Modular<mod>>, bool> = true>
+  static constexpr std::true_type test(Modular<mod>);
+
+  static constexpr std::false_type test(...);
+
+  static constexpr bool result = decltype(test(std::declval<T>()))::value;
+};
+
+}  // namespace modular_internal
+
+template <typename T>
+concept IsModular = modular_internal::IsModularImpl<T>::result;
+
+// Overloads an arithmetic operator with a class method
+#define OVERLOAD_ARITHMETIC_OPERATOR(CONCEPT, OP, METHOD) \
+  template <CONCEPT M>                                    \
+  M operator OP(const M &lhs, const M &rhs) {             \
+    return lhs.METHOD(rhs);                               \
+  }                                                       \
+  template <CONCEPT M, std::convertible_to<M> T>          \
+  M operator OP(const M &lhs, const T &rhs) {             \
+    return lhs.METHOD(static_cast<M>(rhs));               \
+  }                                                       \
+  template <CONCEPT M, std::convertible_to<M> T>          \
+  M operator OP(const T &lhs, const M &rhs) {             \
+    return static_cast<M>(lhs).METHOD(rhs);               \
+  }
+
+// Overloads an inplace operator with a class method
+#define OVERLOAD_INPLACE_OPERATOR(CONCEPT, OP, METHOD) \
+  template <CONCEPT M, std::convertible_to<M> T>       \
+  M &operator OP(M &lhs, const T &rhs) {               \
+    lhs = lhs.METHOD(static_cast<M>(rhs));             \
+    return lhs;                                        \
+  }
+
+// Overloads a pair of comparison operator with a class method
+#define OVERLOAD_COMPARISON_OPERATOR(CONCEPT, OP, OP_NEG, METHOD) \
```

Comment:
Do we really need `OP_NEG` here? We could simply overload both manually. The code for `OP` and `OP_NEG` looks almost the same. This `!` usage can also cause problems for `<` and `>=` case for example. I would prefer manually registering all available operators including negatives.